### PR TITLE
fix(onboarding): rebase when onboardingConfig changes

### LINKED ISSUE #22448: Onboarding branch should update config if onboardingConfig changes and branch is unmodified
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, tell us what version of Renovate you run.

_No response_

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

If an onboarding PR exists, and is unmodified, and its config does not match what is currently in `onboardingConfig`, then Renovate should rebase/repush the onboarding branch.

See #22444 for further details.

To reproduce:
- Run on a repo to create an onboarding PR
- Modify `onboardingConfig` locally
- Run again on the repo

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste the relevant log(s) here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

### PR DESCRIPTION
**Pull request by RahulGautamSingh**
## Changes
Rebase onboarding branch if default onboarding config has changed

-------------------------------
Logic Flow for Re-run on a repo with existing onboarding PR :
--------------------------------
**Before:**
1. If onboardingPr exists we check if it needs to be updated (using cache), and if not needed  we skip processing the branch 
2. but, if an update is required, we check whether the branch is modified or not 
      a) if modified -> check whether its conflicted and if so skip processing the branch
      b) else, merge the base and onboarding branch and then  -> extract -> lookup -> ensureOnboardingPr()
      
**After:**
1. If onboardingPr exists we check if onboarding branch is modifed or not and:
    a) if branch is modifed, skip rebase check and go to step [2]
    b) else, we will check whether the `onboardingConfig` from `config/options` has changed or not; 
           -   if it has changed; rebase the onboarding branch
           -   else continue to step [2];
2. Repeat Steps [1] - [2] from `Before`  (except checking for isModified since its computed)

Creation of onboarding PRs isn't affected by these changes.

## Context
Closes: https://github.com/renovatebot/renovate/issues/22448

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  const contents = await getOnboardingConfigContents(config, configFile);
+
+  if (config.onboardingRebaseCheckbox) {
+    handleOnboardingManualRebase(previousConfigHash, currentConfigHash);
+  }
+
+  if (contents === existingContents) {
+    logger.debug('Onboarding branch is up to date');
+    return null;
+  }
```
Comment (RahulGautamSingh) [inline lib/workers/repository/onboarding/branch/rebase.ts:30]: Should this code be moved to a different function?

Comment (rarkins) [reply lib/workers/repository/onboarding/branch/rebase.ts:30]: Not just for the sake of it. If you have a good reason, then sure

**Comment by RahulGautamSingh** [comment]
Reproduction: https://github.com/RahulGautamSingh-testing/onb-rebase/pull/1

**Comment by RahulGautamSingh** [comment]
Blocked by: https://github.com/renovatebot/renovate/pull/22471

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const contents = await getOnboardingConfigContents(config, configFile);
+  const currentConfigHash = toSha256(contents);
+
+  if (previousConfigHash === currentConfigHash) {
+    logger.debug('Onboarding branch is up to date');
+    return null;
+  }
+
+  logger.debug('Rebasing onboarding branch');
+  // istanbul ignore next
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:25]: seems wrong here

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  const commitMessageFactory = new OnboardingCommitMessageFactory(
+    config,
+    configFile
+  );
+  const commitMessage = commitMessageFactory.create();
+
+  // istanbul ignore if
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:32]: we should add a test, we've had bugs about dry run in the past

**Comment by RahulGautamSingh** [comment]
Blocked for a while:
We don't add conifgHash in comments for onboardingPrs😶, till `onboardingRebaseCheckbox` is set to true. Modifying this old logic.

File: lib/workers/repository/onboarding/branch/index.spec.ts [hunk 4]
```diff
@@ -369,61 +395,46 @@ describe('workers/repository/onboarding/branch/index', () => {
...
 
       it('handles unchecked rebase checkbox', async () => {
         const pr = { bodyStruct: { rebaseRequested: false } };
         platform.getBranchPr.mockResolvedValueOnce(mock<Pr>(pr));
-
         await checkOnboardingBranch(config);
-
         expect(OnboardingState.prUpdateRequested).toBeFalse();
-        expect(git.mergeBranch).toHaveBeenCalledOnce();
-        expect(scm.commitAndPush).toHaveBeenCalledTimes(0);
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.spec.ts:426]: why is this removed? especially the last is a good test to not accidentally commit.

File: lib/workers/repository/onboarding/branch/index.ts [hunk 5]
```diff
@@ -7,7 +7,7 @@ import {
   REPOSITORY_NO_PACKAGE_FILES,
 } from '../../../../constants/error-messages';
 import { logger } from '../../../../logger';
-import type { Pr } from '../../../../modules/platform';
+import { Pr, platform } from '../../../../modules/platform';
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:10]:
```suggestion
import { type Pr, platform } from '../../../../modules/platform';
```

File: lib/workers/repository/onboarding/branch/rebase.spec.ts [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    });
+  });
+
+  describe('rebaseOnboardingBranch()', () => {
+    let config: RenovateConfig;
+    const hash = '';
+
+    beforeEach(() => {
+      memCache.init();
+      jest.resetAllMocks();
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.spec.ts:22]: I think you don't need it, call clearing is set globally

**Review by viceice — COMMENTED** [review]
still blocked?

**Comment by RahulGautamSingh** [comment]
> still blocked?

Yes

File: lib/workers/repository/onboarding/branch/index.spec.ts (hunk 4 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/onboarding/branch/index.spec.ts:426]: forgot to revert after testing ):

**Comment by RahulGautamSingh** [comment]
> still blocked?

Nope

File: lib/workers/repository/onboarding/branch/index.spec.ts [hunk 7]
```diff
@@ -390,23 +416,17 @@ describe('workers/repository/onboarding/branch/index', () => {
...
         platform.getBranchPr.mockResolvedValueOnce(mock<Pr>(pr));
-
         await checkOnboardingBranch(config);
-
         expect(logger.debug).toHaveBeenCalledWith(
           `No rebase checkbox was found in the onboarding PR`
         );
         expect(OnboardingState.prUpdateRequested).toBeTrue();
-        expect(git.mergeBranch).toHaveBeenCalledOnce();
-        expect(scm.commitAndPush).toHaveBeenCalledTimes(0);
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.spec.ts:401]: why removed?

File: lib/workers/repository/onboarding/branch/index.spec.ts [hunk 8]
```diff
@@ -390,23 +416,17 @@ describe('workers/repository/onboarding/branch/index', () => {
       it('detects missing rebase checkbox', async () => {
         const pr = { bodyStruct: { rebaseRequested: undefined } };
         platform.getBranchPr.mockResolvedValueOnce(mock<Pr>(pr));
-
         await checkOnboardingBranch(config);
-
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.spec.ts:395]: please don't remove those newlines, they are for readability

File: lib/workers/repository/onboarding/branch/index.spec.ts (hunk 7 above)
Comment (RahulGautamSingh) [reply lib/workers/repository/onboarding/branch/index.spec.ts:401]: These assertions are extra. We have already checkd with the logger statement that checkbox isn't present.

**Comment by RahulGautamSingh** [comment]
Test-runs:
- onboarding config changed on:
existing pr https://github.com/RahulGautamSingh-testing/old-onb/pull/1
new pr https://github.com/RahulGautamSingh-testing/new-onb/pull/1
- No change to config for 
modified pr https://github.com/RahulGautamSingh-testing/modified-old-onb/pull/1

File: lib/workers/repository/onboarding/branch/index.ts [hunk 9]
```diff
@@ -54,10 +55,29 @@ export async function checkOnboardingBranch(
...
   if (onboardingPr) {
+    logger.debug('Onboarding PR already exists');
+
+    isModified = await isOnboardingBranchModified(config.onboardingBranch!);
+    // if onboarding branch is not modified, check if onboarding config has been changed and rebase if true
+    if (!isModified) {
+      const commit = await rebaseOnboardingBranch(
+        config,
+        onboardingPr.bodyStruct?.rawConfigHash
+      );
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/index.ts:66]: This will constantly rebase if the platform doesn't support embedded html comments in PR's?

Comment (RahulGautamSingh) [reply lib/workers/repository/onboarding/branch/index.ts:66]: Yes, unfortunately.
I think I need to store the configHash in onboardingCache and use it for comaprison.
WDYT?? @rarkins

Comment (viceice) [reply lib/workers/repository/onboarding/branch/index.ts:66]: yes, we should probably ignore this feature there

**Comment by RahulGautamSingh** [comment]
Test-runs on :
existing onboarding pr https://github.com/RahulGautamSingh-testing/old-onbo/pull/1 
new onboarding pr created using this logic https://github.com/RahulGautamSingh-testing/new-onb/pull/2
existing pr that were modified https://github.com/RahulGautamSingh-testing/modified-old-onb/pull/1

**Comment by rarkins** [comment]
Can you please summarize the difference between behavior today vs tomorrow if this is merged?

File: lib/workers/repository/onboarding/branch/rebase.spec.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
+import { RenovateConfig, scm } from '../../../../../test/util';
+import { GlobalConfig } from '../../../../config/global';
+import { logger } from '../../../../logger';
+import * as memCache from '../../../../util/cache/memory';
+import { toSha256 } from '../../../../util/hash';
+import * as _config from './config';
+import { rebaseOnboardingBranch } from './rebase';
+
+const configModule: any = _config;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.spec.ts:9]:
```suggestion
const configModule = mocked(_config);
```

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+export async function rebaseOnboardingBranch(
+  config: RenovateConfig,
+  previousConfigHash: string | undefined
+): Promise<string | null> {
+  logger.debug('Checking if onboarding branch needs rebasing');
+
+  // skip platforms that do not support html comments in pr
+  const pl = GlobalConfig.get('platform')!;
+  if (!['github', 'gitea', 'gitlab'].includes(pl)) {
+    return null;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:19]: @rarkins log not supported debug message?

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+  const commitMessageFactory = new OnboardingCommitMessageFactory(
+    config,
+    configFile
+  );
+  const commitMessage = commitMessageFactory.create();
+
+  if (GlobalConfig.get('dryRun')) {
+    logger.info('DRY-RUN: Would rebase files in onboarding branch');
+    return null;
+  }
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:44]: I think we can save a litle time when doing this before commit message initializing. 🤔

**Comment by RahulGautamSingh** [comment]
> Can you please summarize the difference between behavior today vs tomorrow if this is merged?

Existing onboarding PRs will be rebased if the `onabordingConfig` option changes.

Ie. Currently the default onboarding config only has one field `$schema` and all onabording PR config have this field + preset if available.
If someday another field is added or we decide to update the url in `$schema` ... this feature will make sure that the existing onboarding PRs that have old config gets rebased and re-written with new values, except for the one that have been modified.

This was default behaviour until we decided to move away from rebasing onboarding PRs 
Ref: https://github.com/renovatebot/renovate/pull/22490 https://github.com/renovatebot/renovate/pull/20893

**Comment by rarkins** [comment]
You've not made the today vs tomorrow distinction clear in your description - please edit it or rewrite it so that before/after is clear

**Comment by RahulGautamSingh** [comment]
> You've not made the today vs tomorrow distinction clear in your description - please edit it or rewrite it so that before/after is clear

Updated the description

File: lib/workers/repository/onboarding/branch/rebase.spec.ts [hunk 13]
```diff
@@ -0,0 +1,126 @@
+import { RenovateConfig, mocked, scm } from '../../../../../test/util';
+import { GlobalConfig } from '../../../../config/global';
+import { logger } from '../../../../logger';
+import * as memCache from '../../../../util/cache/memory';
+import { toSha256 } from '../../../../util/hash';
+import * as _config from './config';
+import { rebaseOnboardingBranch } from './rebase';
+
+const configModule: any = mocked(_config);
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.spec.ts:9]:
```suggestion
const configModule = mocked(_config);
```
no any

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+  const commitMessageFactory = new OnboardingCommitMessageFactory(
+    config,
+    configFile
+  );
+  const commitMessage = commitMessageFactory.create();
+
+  // TODO #7154
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:47]: I think you need a different issue id, it's a discussion now

**Review by viceice — COMMENTED** [review]
just a small nit

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
@RahulGautamSingh can you clarify the status of this one. Is it ready/next or should other topics take precedence?

**Comment by RahulGautamSingh** [comment]
It is ready 👍

**Comment by RahulGautamSingh** [comment]
Test-runs:
 When default onboarding config :
1. Un-modified onboarding are rebased https://github.com/RahulGautamSingh-testing/rebase-2/pulls
> after rebasing, on next runs these PRs aren't cloned either
2. Modified onboarding PR isn't rebased nor cloned  https://github.com/RahulGautamSingh-testing/rebase-3/pulls

**Comment by rarkins** [comment]
@RahulGautamSingh I have rewritten this to be a _warning_, not error/abort: #24560

Can you update this PR accordingly?

**Comment by RahulGautamSingh** [comment]
> @RahulGautamSingh I have rewritten this to be a _warning_, not error/abort: #24560
> 
> Can you update this PR accordingly?

Wrong PR 😅

File: lib/workers/repository/onboarding/branch/rebase.ts [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+import { getOnboardingConfigContents } from './config';
+
+export async function rebaseOnboardingBranch(
+  config: RenovateConfig,
+  previousConfigHash: string | undefined,
+): Promise<string | null> {
+  logger.debug('Checking if onboarding branch needs rebasing');
+
+  // skip platforms that do not support html comments in pr
+  const pl = GlobalConfig.get('platform')!;
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.ts:17]: can we use the full name platform for readability?

**Review by viceice — COMMENTED** [review]
otherwise LGTM

**Review by rarkins — APPROVED** [review]

**Pull request merged**